### PR TITLE
registerDisks flags and clickhouse-disks test.

### PR DESCRIPTION
**Pull request by ilejn**
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Disks/ObjectStorages/DiskObjectStorageVFS.cpp [hunk 1]
```diff
@@ -66,7 +66,10 @@ void DiskObjectStorageVFS::startupImpl(ContextPtr context)
     settings.set(std::move(settings_new_version));
 
     if (!enable_gc)
+    {
+        LOG_INFO(log, "VFS GC is not enabled"); /// The log message is used in an integration test
```
Comment (myrrc) [inline src/Disks/ObjectStorages/DiskObjectStorageVFS.cpp:70]:
```suggestion
    {
        LOG_INFO(log, "VFS GC is not enabled");
```

File: src/Disks/registerDisks.h [hunk 2]
```diff
@@ -1,11 +1,37 @@
...
+
 namespace DB
 {
 
-/// @param global_skip_access_check - skip access check regardless regardless
-///                                   .skip_access_check config directive (used
-///                                   for clickhouse-disks)
-void registerDisks(bool global_skip_access_check);
+enum class DiskFlag : size_t
+{
```
Comment (myrrc) [inline src/Disks/registerDisks.h:9]:
```suggestion
enum class DiskFlag
```
No need to specify explicit type, enum is internal to CH

File: src/Disks/registerDisks.h [hunk 3]
```diff
@@ -1,11 +1,37 @@
...
+
+    /// ObjectStorageVFS Garbage Collector is allowed (used for clickhouse-disks)
+    ALLOW_VFS_GC,
+};
+
+/// just a syntax sugar to use scoped enum with bitset without explicit cast
+template <typename FLAG, typename BODY>
+class FlagsSet : public BODY
+{
+    using SelfType = DB::FlagsSet<FLAG, BODY>;
```
Comment (myrrc) [inline src/Disks/registerDisks.h:24]: Use https://github.com/Neargye/magic_enum (Functions for flags)
https://github.com/Neargye/magic_enum/blob/master/example/enum_flag_example.cpp

File: src/Disks/registerDisks.h [hunk 4]
```diff
@@ -1,11 +1,37 @@
...
-///                                   .skip_access_check config directive (used
-///                                   for clickhouse-disks)
-void registerDisks(bool global_skip_access_check);
+enum class DiskFlag : size_t
+{
+    /// skip access check regardless .skip_access_check config directive (used for clickhouse-disks)
+    GLOBAL_SKIP_ACCESS_CHECK = 0,
+
+    /// ObjectStorageVFS is allowed (used e.g. for Keeper)
+    ALLOW_VFS,
```
Comment (myrrc) [inline src/Disks/registerDisks.h:14]:
```suggestion
    ALLOW_VFS = 2,
```

File: src/Disks/registerDisks.h [hunk 5]
```diff
@@ -1,11 +1,37 @@
...
+
+    /// ObjectStorageVFS Garbage Collector is allowed (used for clickhouse-disks)
+    ALLOW_VFS_GC,
```
Comment (myrrc) [inline src/Disks/registerDisks.h:17]:
```suggestion
    ALLOW_VFS_GC = 4,
```

**Review by myrrc — APPROVED** [review]

File: src/Disks/registerDisks.h (hunk 2 above)
Comment (myrrc) [reply src/Disks/registerDisks.h:9]: I also think we could rename it to DiskStartupFlag to resolve ambiguity

File: src/Disks/registerDisks.h (hunk 3 above)
Comment (ilejn) [reply src/Disks/registerDisks.h:24]: CH version of magic_enum is a bit outdated, it does not have enum_containers.
Actually it is the reason why I've introduced FlagsSet.

But Ok, lets use magic_enum.
As far as I understand it requires converting to integer type to check bits (there is no magic here), so we are going to need a function for this.

**Comment by myrrc** [comment]
> CH version of magic_enum is a bit outdated, it does not have 
> enum_containers.

You can update the submodule, a valid point to be used in PR.

**Comment by myrrc** [comment]
Merged in upstream

**Pull request closed without merging**